DynamicTablesPkg/Library/Acpi/Common/AcpiSpmiLib/SpmiGenerator.c

```diff
@@ -0,0 +1,405 @@
+  @param [in]  AcpiTableInfo  Pointer to the ACPI Table Info.
+  @param [in]  CfgMgrProtocol Pointer to the Configuration Manager
+                                Protocol Interface.
+  @param [out] Table          Pointer to the constructed ACPI Table.
+
+  @retval EFI_SUCCESS           Table generated successfully.
+  @retval EFI_INVALID_PARAMETER A parameter is invalid.
+  @retval EFI_NOT_FOUND         The required object was not found.
+  @retval EFI_BAD_BUFFER_SIZE   The size returned by the Configuration
+                                Manager is less than the Object size for the
+                                requested object.
+**/
+STATIC
+EFI_STATUS
+EFIAPI
+BuildSpmiTable (
+  IN  CONST ACPI_TABLE_GENERATOR                  *CONST  This,
+  IN  CONST CM_STD_OBJ_ACPI_TABLE_INFO            *CONST  AcpiTableInfo,
+  IN  CONST EDKII_CONFIGURATION_MANAGER_PROTOCOL  *CONST  CfgMgrProtocol,
+  OUT       EFI_ACPI_DESCRIPTION_HEADER          **CONST  Table
+  )
+{
+  EFI_STATUS                                 Status;
+  CM_ARCH_COMMON_SPMI_INTERFACE_INFO         *SpmiInfo;
+  CM_ARCH_COMMON_SPMI_INTERRUPT_DEVICE_INFO  *SpmiIntrDeviceInfo;
+  IPMI_GET_DEVICE_ID_RESPONSE                DeviceId;
+
+  ASSERT (This != NULL);
+  ASSERT (AcpiTableInfo != NULL);
+  ASSERT (CfgMgrProtocol != NULL);
+  ASSERT (Table != NULL);
+  ASSERT (AcpiTableInfo->TableGeneratorId == This->GeneratorID);
+  ASSERT (AcpiTableInfo->AcpiTableSignature == This->AcpiTableSignature);
+
+  if ((AcpiTableInfo->AcpiTableRevision < This->MinAcpiTableRevision) ||
+      (AcpiTableInfo->AcpiTableRevision > This->AcpiTableRevision))
+  {
+    DEBUG ((
+      DEBUG_ERROR,
+      "ERROR: SPMI: Requested table revision = %d, is not supported."
+      "Supported table revision: Minimum = %d, Maximum = %d\n",
+      AcpiTableInfo->AcpiTableRevision,
+      This->MinAcpiTableRevision,
+      This->AcpiTableRevision
+      ));
+    return EFI_INVALID_PARAMETER;
+  }
+
+  *Table = NULL;
+
+  Status = GetEArchCommonObjSpmiInterfaceInfo (
+             CfgMgrProtocol,
+             CM_NULL_TOKEN,
+             &SpmiInfo,
+             NULL
+             );
+  if (EFI_ERROR (Status)) {
+    DEBUG ((
+      DEBUG_ERROR,
+      "ERROR: SPMI: Failed to retrieve interface type and base address.\n"
+      ));
+    return Status;
+  }
+
+  AcpiSpmi.InterfaceType = SpmiInfo->InterfaceType;
+  CopyMem (
+    &AcpiSpmi.BaseAddress,
+    &SpmiInfo->BaseAddress,
+    sizeof (EFI_ACPI_6_5_GENERIC_ADDRESS_STRUCTURE)
+    );
+
+  Status = GetEArchCommonObjSpmiInterruptDeviceInfo (
+             CfgMgrProtocol,
+             CM_NULL_TOKEN,
+             &SpmiIntrDeviceInfo,
+             NULL
+             );
+  if (EFI_ERROR (Status)) {
+    DEBUG ((
+      DEBUG_INFO,
+      "INFO: SPMI: The platform does not provide interrupt and PCI device information.\n"
+      ));
+    DEBUG ((
+      DEBUG_INFO,
+      "Using default values (0) for the interrupt and device information.\n"

```

Comment:
Shouldn't we abort instead of installing an empty table ?

Reply:
this is intentional, platform configuration might not include the legacy EArchCommonObjSpmiInterruptDeviceInfo.
if not provided then the generator will go with the default value which is zero and valid values as per the specification.